# rubisco-genomics

Code and intermediate data files supporting "Variable impact of geochemical gradients on the functional potential of bacteria, archaea, and phages from the permanently stratified Lac Pavin" (Jaffe et al., 2022).

## code - *.ipynb

Code describing the main analyses and figure generation are available here in Jupyter Notebook format.

```
* LacPavin-binning-curation-public.ipynb - notebook covering genome binning and manual curation 
* LacPavin-diversity-public.ipynb - notebook covering analysis of microbial community composition and distribution
* LacPavin-geo-env.ipynb - notebook covering map-making and plotting of oxygen profiles
* LacPavin-metabolism-public.ipynb - notebook covering analysis of bacterial/archaeal metabolic potential (non-CPR)
* LacPavin-pclust.ipynb - notebook covering analysis of protein clusters for CPR/phage
* LacPavin-phage-analysis-public.ipynb - notebook covering phage distribution, metabolism, diversity, host linkage
* LacPavin-phage-prep.ipynb - notebook covering identification and curation of phage scaffolds
* LacPavin-phylo.ipynb - notebook covering taxonomic assignment and renaming of bins
* LacPavin-rubisco.ipynb - notebook covering analyses of rubisco diversity, distribution, phylogeny
```

Other files or information available upon request to [ajaffe@stanford.edu](mailto:ajaffe@stanford.edu). 
